### ISSUE TITLE
Mocking a mssql DB instance

### ISSUE THREAD

**Issue by jwenz723** [denisenkom/go-mssqldb#486]
Wondering if there is a suggested way of mocking the standard library sql.DB object while making use of this mssql driver?

I tried making use of the datadog [sqlmock](https://github.com/DATA-DOG/go-sqlmock), but it is not working when I pass a TVP type object as a parameter.

Here is an example of my test using sqlmock:
```
type AnyTVP struct{}

// Match satisfies sqlmock.Argument interface
func (a AnyTVP) Match(v driver.Value) bool {
	_, ok := v.(mssql.TVP)
	return ok
}

func Test_InsertContact(t *testing.T) {
	db, mock, err := sqlmock.New()
	if err != nil {
		t.Fatalf("an error '%s' was not expected when opening a stub database connection", err)
	}
	defer db.Close()

	d, err := New(db)
	assert.Nil(t, err)

	contact := Contact{
		ContactID: 123,
	}

	// Setup behavior of mock db instance
	rows := sqlmock.NewRows([]string{"ContactID"})
	rows.AddRow(contact.ContactID)

	mock.ExpectQuery("dbo.InsertContacts").WithArgs(AnyTVP{}).WillReturnRows(rows)

	err = d.InsertContact(context.Background(), contact)
	assert.Nil(t, err)
}
```

When I execute the test I get this error:
```
sql: converting argument with name \"Contacts\" type: unsupported type mssql.TVP, a struct
```

Here is the function that is being called from the test:
```
execTvp = "dbo.InsertContacts @Contacts;"
udtContactTable = "dbo.[UDT_Contact_Table]"

func (s dpservice) InsertContact(ctx context.Context, contact Contact) error {
	tvp := mssql.TVP{
		TypeName: udtContactTable,
		Value:    []Contact{*contact},
	}

	rows, err := s.inCodeDB.QueryContext(ctx, execTvp, sql.Named("Contacts", tvp))
	if err != nil {
		return err
	}
	defer rows.Close()
...
}
```

The reason I am putting this issue here rather than on the sqlmock project is because [this sqlmock issue](https://github.com/DATA-DOG/go-sqlmock/issues/175#issuecomment-500801483) suggests providing a `ValueConverter` when using custom sql drivers. Is there a ValueConverter in this package? I can't find it anywhere?

**Comment by kardianos** [comment]
I wouldn't personally support what you are doing.

My unit tests start by creating a fresh DB and DB schema, populating data in it, then using that fresh DB to run the real queries. Then I test the results of the queries (results, or DB state) for the unit test.

I don't know about sqlmock. Right now I haven't read your issue carefully. But I wanted to point you in the above direction, because over time that is what I found to work best.

**Comment by jwenz723** [comment]
I see what you mean, however, I am just trying to write a simple unit test for my service's behavior around data that is returned from a database stored proc call. I was able to get around my issue in a much simpler way.

I defined an interface that has only the `QueryContext` method since that is all I am using on my sql.DB instance:
```
// QueryContexter is a trimmed down version sql.DB allowing for simpler mocking
type QueryContexter interface {
	QueryContext(ctx context.Context, query string, args ...interface{}) (*sql.Rows, error)
}
```

Then I pass in an instance of `QueryContexter` to my service as a dependency, thus allowing me to create a simple no-op version of `QueryContexter` in my unit tests.

I think your recommendation of running an actual DB instance is a good idea for an integration test. I will probably follow your example to accomplish that as well.
